src/cli/imports.rs

```diff
@@ -0,0 +1,245 @@
+use colored::Colorize;
+use std::fmt;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct Import {
+    name: String,
+    source: String,
+    attrs: Vec<String>,
+    no_color: bool,
+}
+
+impl Import {
+    // Construction helper
+    fn priv_new() -> Import {
+        Import {
+            name: String::from("None"),
+            source: String::from("None"),
+            attrs: vec![],
+            no_color: false,
+        }
+    }
+
+    fn from_elf(
+        elf: &goblin::elf::Elf,
+        sym_idx: &usize,
+        sym: &goblin::elf::sym::Sym,
+        no_color: bool,
+    ) -> Import {
+        fn get_symbol_version_string(elf: &goblin::elf::Elf, sym_idx: &usize) -> Option<String> {
+            let vers_data = elf.versym.as_ref()?.get_at(*sym_idx)?.vs_val;
+
+            if vers_data == 0 {
+                return Some("local".to_string());
+            }
+
+            if let Some(needed) = elf
+                .verneed
+                .as_ref()?
+                .iter()
+                .find(|v| v.iter().any(|f| f.vna_other == vers_data))
+            {
+                if let Some(version) = needed.iter().find(|f| f.vna_other == vers_data) {
+                    let need_str = elf.dynstrtab.get_at(needed.vn_file)?;
+                    let vers_str = elf.dynstrtab.get_at(version.vna_name)?;
+                    return Some(format!("{}, {} ({})", need_str, vers_str, vers_data));
+                }
+            }
+
+            None
+        }
+
+        fn get_got_plt_offset(elf: &goblin::elf::Elf, sym_idx: &usize) -> Option<u64> {
+            let offset = if let Some(reloc) = elf.pltrelocs.iter().find(|r| r.r_sym == *sym_idx) {
+                reloc.r_offset
+            } else {
+                return None;
+            };
+
+            Some(offset)
+        }
+
+        fn get_plt_offset(elf: &goblin::elf::Elf, sym_idx: &usize) -> Option<u64> {
+            let reloc_idx = elf.pltrelocs.iter().position(|r| r.r_sym == *sym_idx)?;
+
+            let mut offset = (reloc_idx as u64 + 1) * 16;
+
+            if let Some(plt) = &elf
```

Comment:
I wrote it this way assuming there was a possibility that the `.plt` section header could be stripped, preventing the function (with its current logic) from locating the `.plt` base address. The section would still exist, so a user may still find it useful to have the calculated offset as output, in case they can find the base `.plt` address another way.

Does that make sense or is my hypothetical case unrealistic?

Reply:
I see! Agree with you then - returning a calculated offset seems preferable to erroring entirely.

Good call on considering stripped binaries. However, I don't *think* the PLT section can be stripped. Not 100% sure, but here's what I learned in attempting to strip it:

* `gcc -s` and `strip --strip-all` both leave a `.plt` section in tact (checking via `readelf --sections <name> | grep plt`), although symbol stripping works as intended. `strip --remove-section=plt` didn't work either.

* Much to my surprise, even compiling a stripped, static binary with `gcc <name>.c -o <name> -s --static -fno-plt` seems to leave `.plt` in there. Per the docs, `-fno-plt` seems [to be an optimization](https://man7.org/linux/man-pages/man1/gcc.1.html) but not a way to actually remove the `.plt` section.

To double check the disassembly (inspired by your other comment), I created a `hello_world` program:

```c
#include <stdio.h>
int main() {
   printf("Hello, ELF World!\n");
   return 0;
}
```

To test the effect of `-fno-plt` in isolation - compiled with `gcc hello.c -o hello`, then ran `gdb -batch -ex "disassemble main" hello`:

```
Dump of assembler code for function main:
   0x0000000000001149 <+0>:     endbr64
   0x000000000000114d <+4>:     push   %rbp
   0x000000000000114e <+5>:     mov    %rsp,%rbp
   0x0000000000001151 <+8>:     lea    0xeac(%rip),%rdi        # 0x2004
   0x0000000000001158 <+15>:    callq  0x1050 <puts@plt>
   0x000000000000115d <+20>:    mov    $0x0,%eax
   0x0000000000001162 <+25>:    pop    %rbp
   0x0000000000001163 <+26>:    retq
End of assembler dump.
```

Doing the same but adding `-fno-plt` to the `gcc` params:

```
Dump of assembler code for function main:
   0x0000000000001129 <+0>:     endbr64
   0x000000000000112d <+4>:     push   %rbp
   0x000000000000112e <+5>:     mov    %rsp,%rbp
   0x0000000000001131 <+8>:     lea    0xecc(%rip),%rdi        # 0x2004
   0x0000000000001138 <+15>:    callq  *0x2e9a(%rip)        # 0x3fd8
   0x000000000000113e <+21>:    mov    $0x0,%eax
   0x0000000000001143 <+26>:    pop    %rbp
   0x0000000000001144 <+27>:    retq
End of assembler dump.
```

So `-fno-plt` does indeed inline call resolution.

In the latter case, the `--imports` output is empty (which is fine, GOT stuff got inlined directly, it would be pretty crazy to have to resolve such an optimization even if technically possible!). BUT there is still a `.plt` section in the binary.

So, in conclusion: I couldn't strip the PLT no matter how hard I tried. Learned that the `-fno-plt` flag doesn't actually mean "no plt", although it does create another edge case (don't worry about it, seriously - the ELF format has endless edge cases).

Reply:
Awesome, thanks for looking into this! I took your advice and updated the `get_plt_address` function to return `None` if the `.plt` lookup fails. I also explicitly ensure we are dealing with a `JUMP_SLOT` relocation to try to cut out a few more of those endless edge cases.

https://github.com/d3vco/xgadget/blob/ac67ffc7ace35f35fb3a7c14056fd1da22427f72/src/cli/imports.rs#L402-L419